src/atlas/redistribution/detail/RedistributeGeneric.cc

```diff
@@ -0,0 +1,447 @@
+/*
+ * (C) Crown Copyright 2021 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#include <numeric>
+#include <vector>
+
+#include "atlas/field/Field.h"
+#include "atlas/functionspace/CellColumns.h"
+#include "atlas/functionspace/EdgeColumns.h"
+#include "atlas/functionspace/NodeColumns.h"
+#include "atlas/functionspace/PointCloud.h"
+#include "atlas/functionspace/StructuredColumns.h"
+#include "atlas/parallel/mpi/mpi.h"
+#include "atlas/redistribution/detail/RedistributionImplFactory.h"
+#include "atlas/redistribution/detail/RedistributeGeneric.h"
+#include "atlas/util/Unique.h"
+
+
+namespace atlas {
+namespace redistribution {
+namespace detail {
+
+using mesh::HybridElements;
+using mesh::Nodes;
+
+// Helper type definitions and functions for redistribution.
+namespace  {
+
+// Define index-UID struct. (Needed to overload "<").
+struct IdxUid : public std::pair<idx_t, uidx_t> {
+    using std::pair<idx_t, uidx_t>::pair;
+};
+
+Field getGhostField( const FunctionSpace& functionspace ) {
+    if( functionspace::NodeColumns( functionspace ) ||
+        functionspace::StructuredColumns( functionspace ) ||
+        functionspace::PointCloud( functionspace ) ) {
+        return functionspace.ghost();
+    }
+    if( functionspace::EdgeColumns( functionspace ) ||
+        functionspace::CellColumns( functionspace ) ) {
+
```

Comment:
Good old SFINAE. Clever, but often pointless!

Reply:
It has its purpose, but was probably overkill here, and we can work with the handles instead.

Reply:
Oh I agree. And what you did fits much better with other polymorphism patterns in Atlas.

Still, I enjoyed my moment as a mad sfinae scientist! 😂